keps/sig-api-machinery/20190426-crd-defaulting.md

```diff
@@ -0,0 +1,306 @@
+
+Mutating admission webhooks can be used to partially replicate defaulting for incoming request payloads. But mutating admission webhooks do not run when reading from etcd.
+
+This KEP is about adding support for specifying default values for fields via OpenAPI v3 validation schemas in the CRD manifest. OpenAPI v3 has native support for a `default` field with arbitrary JSON values, for example: 
+
+```yaml
+properties:
+  foo:
+    type: string
+    default: "abc" 
+```
+
+This KEP proposes to apply these default values during deserialization, in the same way as native resources do. We assume _structural schemas_ as defined in [KEP Vanilla OpenAPI Subset: Structural Schema](/keps/sig-api-machinery/20190425-structural-openapi.md).
+
+This feature starts behind a feature gate `CustomResourceDefaulting`, disabled by default as alpha.
+
+## Motivation
+
+* By far most native Golang based resources implement defaulting. CRD do not allow that, leading to unnatural, not Kubernetes-like APIs. This is bad for the ecosystem.
+* Mutating Admission Webhooks can be used for defaulting, but:
+  * they are not adequat as it is not possible to set default values of newly added fields on GET because admission is not run in the storage layer when reading from etcd.
+  * webhooks are complex, both from the development/maintenance point of view and due to their non-trivial operational overhead. This makes them a no-go for many "not so ambitiously, professionally developed CRDs", e.g. in in-house enterprise environments.
+* _Structural schemas_ as defined in [KEP Vanilla OpenAPI Subset: Structural Schema](/keps/sig-api-machinery/20190425-structural-openapi.md) makes defaulting a low hanging fruit.
+ 
+### Goals
+
+* add CustomResource defaulting at the correct position in the request pipeline 
+* allow definition of defaults via the OpenAPI v3 `default` field.
+
+### Non-Goals
+
+* allow non-constant defaults: native Golang code can of course set defaults which depends on other fields of a JSON object. This is out of scope here and would need some kind of defaulting webhook.
+* native-type declarative defaulting: this KEP is about CRDs. Though it might be desirable to support the same `// +default=<some-json-value>` tag and a mechanism in `k8s.io/apiserver` to evaluate defaults in native, generic registries, this is out-of-scope of the KEP though.
+
+## Proposal
+
+We assume the CRD has _structural schemas_ (as defined in [KEP Vanilla OpenAPI Subset: Structural Schema](/keps/sig-api-machinery/20190425-structural-openapi.md)).
+
+We propose to
+
+1. derive the value-validation-less variant of the structural schema (trivial by definition of _structural schema_) and 
+2. recursively follow the given CustomResource instance and the structural schema, applying defaults where an object field is 
+  * undefined (`_, ok := obj[field]; !ok`)
+  * `nil` if the field not nullable
+  * empty in case of lists and maps, and if nullable is not set.
```

Comment:
> in case of lists

clarify how we determine this. `type: "array"` in the schema? `typeof(default) == []interface{}`? typeof(value) == []interface{}`? all of the above?

> and maps

clarify how we determine this. `type: "object"` in the schema? `typeof(default) == map[string]interface{}`? typeof(value) == map[string]interface{}`? all of the above?

I mostly want to make sure defaulting doesn't replace an empty array or map with a default value of the correct type prior to validation and mask what should be reported to the user as a schema validation error

Reply:
Made the cases explicit as you wrote down above.

Note: we could do type validation during pruning, i.e. as part of the deserialization process, to match native types.